### DISCUSSION TITLE
Increasing quality of video streams

### DISCUSSION THREAD

**Discussion by wolfgang55283** [savonet/liquidsoap#4046] [Q&A]
Hey,

I started using the liquidsoap software on an ubuntu server some days ago. Streaming audio playlists works perfectly well. Today I wanted to stream a single video file. It worked, too but the video is quite blurry (low quality) and a little bit sticky. Just the video - the audio is the same quality as when I watch it local on a player)

So I tried to set the bitrate and the frames but liquidsoap said that these operations are unknow. Well, I must admit that I am not an expert in using ffmpeg, so maybe you can help me increasing the quality of the video stream or tell me if the reason for the low quality is something else.

Here is my script in order to give you information you may need:

`source = single("test.mp4")

output.icecast(
    %ffmpeg(
        format="ogg",
        %audio(codec="libopus", samplerate=48000),
        %video(codec="libtheora")
    ),
    host = "localhost",
    port = 5555,
    password = "topsecret",
    mount = "/stream",
    source
)`

Would be happy to get some hints. Thanks in advance.

**Comment by gAlleb** [comment]
Hi @wolfgang55283! There are numerous possibilities with video setup.
You can read about them here https://www.liquidsoap.info/doc-dev/ffmpeg_encoder.html and also in LS book http://www.liquidsoap.info/book/book.pdf.

There is even an option to use `theora` outside ffmpeg like `output.icecast(%ogg(%opus, %theora)` You can see it on p. 226 of the book. Other possible options can also be found there - p. 223.

Regarding your question, you may want to specify the bitrate with `b` parameter. E.g.:

```ruby
output.icecast(
%ffmpeg(
format="ogg",
%audio(codec="libopus", samplerate=48000),

%video(codec="libtheora", b=10000000)

),
```

And maybe other parameters like GOP (`g=30`) or other combinations (e.g. `global_quality=5`). E.g. `q=7` sets VBR and `b` option (CBR) is ignored then. You can find more about them on ffmpeg site -> https://ffmpeg.org/ffmpeg-codecs.html and play with them till you get the desired result.

The quality and stickiness of the video will be dependent on your CPU. So if it's possible and you have access to it - you can try Hardware Acceleration.

Personally I think `mpegts` with `libx264` or `libx265` are better somehow. Below you can see one of my video setups. It's icecast compatible. Don't forget to add `send_icy_metadata=true`.

```ruby
(%ffmpeg(format="mpegts",
  %audio(
    codec="aac",
    samplerate=44100,
    channels=2,
    b="192k",
    profile="aac_low"
  ),
  %video(codec="libx264",
    g="50",
    maxrate="8000k",
    bufsize="16000k",
    preset="ultrafast",  
    tune="zerolatency")), send_icy_metadata=true, ...
```

**Reply by wolfgang55283** [reply]
Thanks a lot for your very detailed information. I will try the hints you gave me and read the important sections in the documentation.
If none of this should heldp, I think the cpu or the memory of the server is too low.
Thanks again, I really appreciate your help
